securestorage2/src/main/java/de/adorsys/android/securestorage2/internal/KeyStoreToolApi21.kt

```diff
@@ -0,0 +1,318 @@
+        )
+
+        return secretKeys
+    }
+
+    private fun storeAesKeyPart(
+        context: Context,
+        keyStoreInstance: KeyStore,
+        cipher: Cipher,
+        keyValueKey: String,
+        aesKey: String
+    ) {
+        val publicKey = getPublicKey(context, keyStoreInstance)
+
+        cipher.init(Cipher.ENCRYPT_MODE, publicKey)
+
+        val bytes = cipher.doFinal(aesKey.toByteArray())
+
+        SecureStorage.getSharedPreferencesInstance(context).edit()
+            .putString(keyValueKey, Base64.encodeToString(bytes, Base64.DEFAULT))
+            .execute(SecureStorageConfig.INSTANCE.ASYNC_OPERATION)
+    }
+
+    private fun getAesKeyPart(
+        context: Context,
+        keyStoreInstance: KeyStore,
+        cipher: Cipher,
+        keyValueKey: String
+    ): String {
+        val privateKey = getPrivateKey(context, keyStoreInstance)
+
+        cipher.init(Cipher.DECRYPT_MODE, privateKey)
+
+        val encodedKey = SecureStorage.getSharedPreferencesInstance(context).getString(keyValueKey, null)
+
+        if (encodedKey == null) {
+            throw SecureStorageException(
+                context.getString(R.string.message_key_does_not_exist),
+                null,
+                SecureStorageException.ExceptionType.KEYSTORE_EXCEPTION
+            )
+        } else {
+            val decodedKey = Base64.decode(encodedKey, Base64.DEFAULT)
+
+            val decryptedKey = cipher.doFinal(decodedKey)
+            return String(decryptedKey)
+        }
+    }
+
+    private fun encodeKeyToString(key: SecretKey): String {
+        return Base64.encodeToString(key.encoded, Base64.DEFAULT)
+    }
+
+    private fun decodeStringToKey(encodedString: String): SecretKey {
+        val encodedKey = Base64.decode(encodedString, Base64.DEFAULT)
+        return SecretKeySpec(encodedKey, 0, encodedKey.size, AES_ALGORITHM)
+    }
+
+    private fun getAesKey(context: Context, keyStoreInstance: KeyStore, cipher: Cipher, keyPrefix: String): SecretKeys {
+        val encodedIntegrityKey = getAesKeyPart(context, keyStoreInstance, cipher, keyPrefix + KEY_AES_INTEGRITY_KEY)
+        val encodedConfidentialityKey =
+            getAesKeyPart(context, keyStoreInstance, cipher, keyPrefix + KEY_AES_CONFIDENTIALITY_KEY)
+
+        val reconstructedIntegrityKey = decodeStringToKey(encodedIntegrityKey)
+        val reconstructedConfidentialityKey = decodeStringToKey(encodedConfidentialityKey)
+
+        return SecretKeys(reconstructedConfidentialityKey, reconstructedIntegrityKey)
+    }
+
+    @Throws(SecureStorageException::class)
+    private fun getPrivateKey(context: Context, keyStoreInstance: KeyStore): PrivateKey {
+        val privateKey: PrivateKey
+
+        keyStoreInstance.getKey(SecureStorageConfig.INSTANCE.ENCRYPTION_KEY_ALIAS, null)
+        try {
+            if (rsaKeyPairExists(keyStoreInstance)) {
+                privateKey = if (VERSION.SDK_INT >= VERSION_CODES.P) {
```

Comment:
instead of assigning the value to `privateKey` return the value of the `when` expression directly like `return when(VERSION.SDK_INT >= VERSION_CODES.P) { true -> keyStoreInstance.getKey(SecureStorageConfig.INSTANCE.ENCRYPTION_KEY_ALIAS, null) as PrivateKey; false ->  privateKeyEntry.privateKey }`